// @soulmachines/smwebsdk/lib/LocalSession.js
"use strict";
/**
 * @module smwebsdk
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalSession = void 0;
var tslib_1 = require("tslib");
/*
 * Copyright 2017-2020 Soul Machines Ltd. All Rights Reserved.
 */
var Deferred_1 = require("./Deferred");
var Features_1 = require("./Features");
var Logger_1 = require("./utils/Logger");
var make_error_1 = require("./utils/make-error");
/**
 *  LocalSession class
 */
var LocalSession = /** @class */ (function () {
    function LocalSession(videoElement, logger) {
        if (logger === void 0) { logger = new Logger_1.Logger(); }
        var _this = this;
        this.logger = logger;
        this._isMicrophoneConnected = false;
        this._isCameraConnected = false;
        this._onConnectedStorage = function (resumeRequested, isResumedSession, server, sessionId
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        ) { };
        this._closed = false;
        this._outgoingQueue = [];
        // Duration that microphone mute is maintained by the web sdk after the persona has
        // finished speaking.  Set to -1 to disable.  Default value is -1 (disabled).
        this._microphoneMuteDelay = -1;
        this._offsetX = 0;
        this._offsetY = 0;
        if (videoElement) {
            this._viewport_element = videoElement;
        }
        window.SmRuntimeHostReceiveMessage = this.receiveMessage.bind(this);
        if (typeof window.SmRuntimeHostStyleViewportElement === 'function') {
            window.SmRuntimeHostStyleViewportElement(this._viewport_element);
        }
        // owner specifies custom close method
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        this._onClose = function (reason) { };
        // owner specifies custom message handler
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        this._onMessage = function (message) { };
        // owner specifies custom rtc user text message handler
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        this._onUserText = function (text) { };
        this.sendVideoBounds(0, 0);
        // The initial positioning can take a while. Would be nice to make this more deterministic.
        setTimeout(function () {
            _this.sendVideoBounds(0, 0);
        }, 3000);
        this._features = new Features_1.Features();
        this.log('Local session created!');
    }
    LocalSession.prototype.receiveMessage = function (raw_text) {
        var _a, _b;
        var message = JSON.parse(raw_text);
        this.log("message received: ".concat(raw_text));
        this._onMessage(message);
        if (message.name === 'state' &&
            message.category === 'scene' &&
            ((_b = (_a = message.body) === null || _a === void 0 ? void 0 : _a.session) === null || _b === void 0 ? void 0 : _b.state) === 'idle') {
            this.log('Local session ending - conversationEnded');
            this.close(true, 'conversationEnded');
        }
    };
    Object.defineProperty(LocalSession.prototype, "onConnected", {
        set: function (sessionFunction) {
            this._onConnectedStorage = sessionFunction;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "onClose", {
        set: function (closeFunction) {
            this._onClose = closeFunction;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "onMessage", {
        set: function (messageFunction) {
            this._onMessage = messageFunction;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "onUserText", {
        set: function (userTextFunction) {
            this._onUserText = userTextFunction;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "loggingEnabled", {
        get: function () {
            return this.logger.isEnabled;
        },
        /**
         * @deprecated use setLogging(boolean).
         */
        set: function (enable) {
            this.logger.log('warn', 'loggingEnabled is deprecated and will be removed in a future version. Please use setLogging(boolean)');
            this.logger.enableLogging(enable);
        },
        enumerable: false,
        configurable: true
    });
    LocalSession.prototype.setMinLogLevel = function (level) {
        this.logger.setMinLogLevel(level);
    };
    LocalSession.prototype.setLogging = function (enable) {
        this.logger.enableLogging(enable);
    };
    LocalSession.prototype.log = function (text) {
        this.logger.log('log', text);
    };
    LocalSession.prototype.sendVideoBounds = function (widthIgnored, heightIgnored) {
        var _this = this;
        // We need to defer the update very slightly to give the browser time to reflow,
        // otherwise we get out of date values for width, height etc:
        setTimeout(function () {
            // Brute-force method for getting pos and dimensions, as
            // getBoundingClientRect seems to be unreliable (sometimes
            // returning zeroes for left and right):
            var el = _this._viewport_element;
            if (el) {
                var view = document.defaultView || window;
                var width = parseInt(view.getComputedStyle(el).width, 10);
                var height = parseInt(view.getComputedStyle(el).height, 10);
                _this._offsetX = 0;
                _this._offsetY = 0;
                while (el && !isNaN(el.offsetLeft) && !isNaN(el.offsetTop)) {
                    _this._offsetX += el.offsetLeft - el.scrollLeft;
                    _this._offsetY += el.offsetTop - el.scrollTop;
                    el = el.offsetParent;
                }
                if (document.documentElement) {
                    var x_off = document.documentElement.scrollLeft;
                    var y_off = document.documentElement.scrollTop;
                    _this._offsetX -= x_off;
                    _this._offsetY -= y_off;
                }
                _this.log("Updating bounds: x =  ".concat(_this._offsetX, " , y = ").concat(_this._offsetY, "', w = ").concat(width, ", h = ").concat(height));
                // update bounds
                var top_1 = _this._offsetY;
                var left = _this._offsetX;
                var bottom = _this._offsetY + height;
                var right = _this._offsetX + width;
                var payload = {
                    name: 'videoBounds',
                    body: { top: top_1, left: left, bottom: bottom, right: right },
                    category: 'local',
                    kind: 'event',
                };
                _this.sendMessage(payload);
            }
        }, 0);
    };
    LocalSession.prototype.hideVideo = function () {
        var top = 0;
        var left = 0;
        var bottom = 0;
        var right = 0;
        var payload = {
            name: 'videoBounds',
            body: { top: top, left: left, bottom: bottom, right: right },
            category: 'local',
            kind: 'event',
        };
        this.sendMessage(payload);
    };
    LocalSession.prototype.sendRtcEvent = function (name, body) {
        // NOOP: Stuff for compatibility with Session in Scene
    };
    LocalSession.prototype.connect = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var deferred, result, payload;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        deferred = new Deferred_1.Deferred();
                        this.log('Local session connecting!');
                        this._closed = false;
                        return [4 /*yield*/, this._features.detectWebRTCFeatures()];
                    case 1:
                        result = _a.sent();
                        this._closed = false;
                        this._sessionId = undefined;
                        this._isMicrophoneConnected = result.hasMicrophone;
                        this._isCameraConnected = result.hasCamera;
                        if (typeof window.local_websocket_port === 'number') {
                            this._serverConnection = new WebSocket('ws://localhost:' + window.local_websocket_port);
                            this.log('websocket open');
                            this._serverConnection.onmessage = function (msg) {
                                _this.gotMessageFromServer(msg);
                            };
                            this._serverConnection.onerror = function (event) {
                                if (deferred.isPending()) {
                                    deferred.reject((0, make_error_1.makeError)('websocket failed', 'serverConnectionFailed'));
                                }
                            };
                            this._serverConnection.onopen = function (event) {
                                // disable SmRuntimeHostReceiveMessage
                                // eslint-disable-next-line @typescript-eslint/no-empty-function
                                window.SmRuntimeHostReceiveMessage = function () { };
                                _this.log('Local session connected!');
                                // send out messages in queue
                                for (var i = 0; i < _this._outgoingQueue.length; i++) {
                                    _this._serverConnection.send(JSON.stringify(_this._outgoingQueue[i]));
                                    _this.logger.log('log', 'SmLocalSession.prototype.sendMessage, forwarding message to Web Socket: ' +
                                        _this._outgoingQueue[i]);
                                }
                                _this._outgoingQueue = [];
                                if (deferred.isPending()) {
                                    deferred.resolve();
                                }
                            };
                            this._serverConnection.onclose = function (event) {
                                _this.logger.log('log', "websocket closed: code(".concat(event.code, "), reason(").concat(event.reason, "), clean(").concat(event.wasClean, ")"));
                                if (!deferred.isRejected) {
                                    _this.close(false, 'normal');
                                }
                            };
                        }
                        else {
                            this.log('local_websocket_port not found! Failed to create WebSocket');
                            if (deferred.isPending()) {
                                deferred.reject((0, make_error_1.makeError)('websocket failed', 'local_websocket_port not found'));
                            }
                        }
                        payload = {
                            name: 'startSession',
                            body: {},
                            category: 'scene',
                            kind: 'request',
                        };
                        this.sendMessage(payload);
                        return [2 /*return*/, deferred.promise];
                }
            });
        });
    };
    LocalSession.prototype.gotMessageFromServer = function (websocket_message) {
        var raw_text = websocket_message.data;
        var message = JSON.parse(raw_text);
        var category = message.category;
        var name = message.name;
        var body = message.body;
        if (category !== 'webrtc') {
            // forward on non-webrtc messages (e.g. scene)
            this._onMessage(message);
        }
        else if (name === 'close') {
            this.close(false, body.reason);
        }
        if (name === 'state' &&
            category === 'scene' &&
            body.session !== null &&
            body.session !== undefined &&
            body.session.state === 'idle') {
            this.log('Local session ending due to server idle message');
            this.close(true, 'conversationEnded');
        }
    };
    LocalSession.prototype.sendMessage = function (message) {
        var msg = JSON.stringify(message);
        if (this._serverConnection &&
            this._serverConnection.readyState === WebSocket.OPEN) {
            this._serverConnection.send(msg);
            this.log("SmLocalSession.prototype.sendMessage, forwarding message to Web Socket: ".concat(msg));
        }
        else {
            this._outgoingQueue.push(message);
        }
    };
    LocalSession.prototype.sendUserText = function (text) {
        this.logger.log('log', 'SmLocalSession.prototype.sendUserText, discarding text: ' + text);
    };
    LocalSession.prototype.close = function (sendRtcClose, reason) {
        if (sendRtcClose === void 0) { sendRtcClose = true; }
        if (reason === void 0) { reason = 'normal'; }
        if (this._closed) {
            return;
        }
        this._closed = true;
        this._onClose(reason);
        this._isMicrophoneConnected = false;
        this._isCameraConnected = false;
        this.hideVideo();
        if (this._serverConnection) {
            this.log('closing server connection');
            var normalClosureCode = 1000;
            this._serverConnection.close(normalClosureCode, reason);
        }
    };
    Object.defineProperty(LocalSession.prototype, "peerConnection", {
        get: function () {
            return null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "userMediaStream", {
        get: function () {
            return null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "serverConnection", {
        get: function () {
            return this._serverConnection;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "sessionId", {
        get: function () {
            return this._sessionId;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "isMicrophoneConnected", {
        get: function () {
            return this._isMicrophoneConnected;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "isCameraConnected", {
        get: function () {
            return this._isCameraConnected;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "features", {
        get: function () {
            return this._features;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "microphoneMuteDelay", {
        get: function () {
            return this._microphoneMuteDelay;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "microphoneMuted", {
        get: function () {
            // todo - RuntimeHost does not yet support this,
            //        currently only needed in webrtc sessions and tests
            if (typeof window.SmRuntimeHostIsMicrophoneMuted === 'function') {
                return window.SmRuntimeHostIsMicrophoneMuted();
            }
            return false;
        },
        set: function (mute) {
            // todo - RuntimeHost does not yet support this,
            //        currently only needed in webrtc sessions and tests
            if (typeof window.SmRuntimeHostMuteMicrophone === 'function') {
                window.SmRuntimeHostMuteMicrophone(mute);
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "offsetX", {
        get: function () {
            return this._offsetX;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(LocalSession.prototype, "offsetY", {
        get: function () {
            return this._offsetY;
        },
        enumerable: false,
        configurable: true
    });
    LocalSession.prototype.isMicrophoneActive = function () {
        return this.isMicrophoneConnected && !this.microphoneMuted;
    };
    LocalSession.prototype.isCameraActive = function () {
        return this.isCameraConnected;
    };
    LocalSession.prototype.setMediaDeviceActive = function (_a) {
        var microphone = _a.microphone, camera = _a.camera;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_b) {
                throw (0, make_error_1.makeError)('setMediaDeviceActive not supported on LocalSession', 'notSupported');
            });
        });
    };
    return LocalSession;
}());
exports.LocalSession = LocalSession;
