// @soulmachines/smwebsdk/lib-esm/Persona.js
/**
 * @module smwebsdk
 */
import { SmEvent } from './SmEvent';
/**
 * Persona class to control a scene persona
 * @public
 */
var Persona = /** @class */ (function () {
    /**
     * @param scene - The scene the persona resides in
     * @param personaId - The id of the persona as received in a state callback
     */
    function Persona(scene, personaId) {
        this._scene = scene;
        this._personaId = personaId;
        if (!this._scene.onConversationResultEvents[personaId]) {
            this._scene.onConversationResultEvents[personaId] = new SmEvent();
        }
        // /** Event which will be triggered whenever conversation results are received. Use
        //   *  onConversationResultEvents.addListener() to register a callback for this event.
        //   *  The single parameter to the callback will be an object with the fields:
        //   *    - **status**        - Status code
        //   *    - **errorMessage**  - Optional error strubg
        //   *    - **personaId**     - Numeric ID of Persona
        //   *    - **output:**
        //   *      - **text**        - Textual response from the converstation backend
        //   *    - **context**       - Dictionary of public conversation variables
        //   *    - **provider:**
        //   *      - **kind**        - Conversation backend name, eg "watson"
        //   *      - **meta**        - Conversation backend specific metadata
        //   */
        this._onConversationResultEvent =
            this._scene.onConversationResultEvents[personaId];
        if (!this._scene.onSpeechMarkerEvents[personaId]) {
            this._scene.onSpeechMarkerEvents[personaId] = new SmEvent();
        }
        // /** Event which will be triggered whenever a speech marker is reached. Use
        //   * onSpeechMarkerEvents.addListener() to register a callback for this event.
        //   * A speech marker is triggered using the format @marker(markername, param0,... paramn)
        //   * in speech. There are also other verbs which establish speech markers.
        //   * Eg @showcards(cardA, cardB). When using a speech marker with Watson Assistant,
        //   * the @ symbol must be escaped with a backslash, as follows:
        //   * \@marker(markername, param0,... paramn)
        //   * The single parameter to the callback will be an object with the fields:
        //   *    - **personaId**     - Numeric ID of Persona
        //   *    - **name**          - Kind of speech marker, eg, "showcards", "hidecards"
        //   *    - **arguements**    - Marker specific arguements.
        //   */
        this._onSpeechMarkerEvent = this._scene.onSpeechMarkerEvents[personaId];
    }
    /**
     * Start speaking the given text
     * @param text - The text to speak
     * @param context - The context included in the transcript
     * @param optionalArgs - Optional start speaking arguments or null
     */
    Persona.prototype.startSpeaking = function (text, context, optionalArgs) {
        if (context === void 0) { context = null; }
        if (optionalArgs === void 0) { optionalArgs = null; }
        var body = {
            personaId: this._personaId,
            text: text,
        };
        if (context) {
            body.context = context;
        }
        if (optionalArgs) {
            body.optionalArgs = optionalArgs;
        }
        return this._scene.sendRequest('startSpeaking', body);
    };
    /**
     * Stop speaking
     */
    Persona.prototype.stopSpeaking = function () {
        var body = {
            personaId: this._personaId,
        };
        return this._scene.sendRequest('stopSpeaking', body);
    };
    /**
     * Send a chat message to conversation
     * @param text - Text to send to conversation
     * @param variables - Variables to send to the conversation provider
     * @param optionalArgs - Optional arguments object (none currently supported)
     */
    Persona.prototype.conversationSend = function (text, variables, optionalArgs) {
        var body = {
            personaId: this._personaId,
            text: text,
            variables: variables,
            optionalArgs: optionalArgs,
        };
        return this._scene.sendRequest('conversationSend', body);
    };
    /**
     * Set variables that will be applied on the next conversation request
     * triggered by speech to text or a conversationSend()
     * @param variables - Variables to send to the conversation provider
     */
    Persona.prototype.conversationSetVariables = function (variables) {
        var body = {
            personaId: this._personaId,
            variables: variables,
        };
        return this._scene.sendRequest('conversationSetVariables', body);
    };
    /**
     * Get variables will return the current conversation variable values
     * in the promise completion.
     * Completion/promise receives an object with member 'variables'.
     */
    Persona.prototype.conversationGetVariables = function () {
        var body = {
            personaId: this._personaId,
        };
        return this._scene.sendRequest('conversationGetVariables', body);
    };
    /**
     * Cut or animate to a named camera position, with support for camera adjustment.
     * @param cameraName - Named camera position. The currently supported option is "CloseUp".
     * @param time - Time in seconds for the animation to run. 0 indicates a cut.
     * @param orbitDegX - Degrees of horizontal rotation around implicit orbit point of camera position (typically the middle of the Persona's head)
     * @param orbitDegY - Degrees of vertical rotation around implicit orbit point of camera position (typically the middle of the Persona's head)
     * @param panDeg - Degrees of camera pan. Orbital adjustment is applied before pan adjustment.
     * @param tiltDeg - Degrees of camera tilt. Orbital adjustment is applied before tilt adjustment.
     */
    Persona.prototype.animateToNamedCameraWithOrbitPan = function (cameraName, time, orbitDegX, orbitDegY, panDeg, tiltDeg) {
        var body = {
            personaId: this._personaId,
            cameraName: cameraName,
            time: time,
            orbitDegX: orbitDegX,
            orbitDegY: orbitDegY,
            panDeg: panDeg,
            tiltDeg: tiltDeg,
        };
        return this._scene.sendRequest('animateToNamedCamera', body);
    };
    /**
     * Play an animation.
     * @internal
     * @param animation - Structured animation data.
     */
    Persona.prototype.playAnimation = function (animation) {
        var body = {
            personaId: this._personaId,
            animation: animation,
        };
        return this._scene.sendRequest('playAnimation', body);
    };
    /**
     * Get & Set bl variables.
     * @internal
     */
    Persona.prototype.getVariables = function (names, errorTolerant, format) {
        if (errorTolerant === void 0) { errorTolerant = false; }
        if (format === void 0) { format = ''; }
        var body = {
            personaId: this._personaId,
            names: names,
            errorTolerant: errorTolerant,
            format: format,
        };
        return this._scene.sendRequest('getVariables', body);
    };
    /**
     * @internal
     */
    Persona.prototype.setVariables = function (variables) {
        var body = {
            personaId: this._personaId,
            Variables: variables,
        };
        return this._scene.sendRequest('setVariables', body);
    };
    /**
     * @internal
     */
    Persona.prototype.setVariablesOneway = function (variables) {
        var body = {
            personaId: this._personaId,
            Variables: variables,
        };
        this._scene.sendOnewaySceneRequest('setVariables', body);
    };
    /**
     * Get bl variables list.
     * @internal
     */
    Persona.prototype.getVariablesList = function () {
        var body = {
            personaId: this._personaId,
        };
        return this._scene.sendRequest('getVariablesList', body);
    };
    /**
     * Get model bl variables list.
     * @internal
     */
    Persona.prototype.getModelVariablesList = function (modelName) {
        var body = {
            personaId: this._personaId,
            Models: modelName,
        };
        return this._scene.sendRequest('getModelVariablesList', body);
    };
    /**
     * Get model children.
     * @internal
     */
    Persona.prototype.getModelChildren = function (modelName) {
        var body = {
            personaId: this._personaId,
            Models: modelName,
        };
        return this._scene.sendRequest('getModelChildren', body);
    };
    /**
     * Get model list by snippet.
     * @internal
     */
    Persona.prototype.getModelFilterSearchResult = function (modelName) {
        var body = {
            personaId: this._personaId,
            Models: modelName,
        };
        return this._scene.sendRequest('getModelFilterSearchResult', body);
    };
    /**
     * Get model variable list by snippet.
     * @internal
     */
    Persona.prototype.getModelVariableFilterSearchResult = function (variableName) {
        var body = {
            personaId: this._personaId,
            Models: variableName,
        };
        return this._scene.sendRequest('getModelVariableFilterSearchResult', body);
    };
    /**
     * Get connector entries.
     * @internal
     */
    Persona.prototype.getConnectorEntries = function (model) {
        var body = {
            personaId: this._personaId,
            model: model,
        };
        return this._scene.sendRequest('getConnectorEntries', body);
    };
    /**
     * Start BL profiling.
     * @internal
     */
    Persona.prototype.startBlProfiling = function () {
        var body = {
            personaId: this._personaId,
        };
        return this._scene.sendRequest('startBlProfiling', body);
    };
    /**
     * Stop BL profiling.
     * @internal
     */
    Persona.prototype.stopBlProfiling = function (reverse) {
        var body = {
            personaId: this._personaId,
            reverse: reverse,
        };
        return this._scene.sendRequest('stopBlProfiling', body);
    };
    /**
     * Get model hierarchy.
     * @internal
     */
    Persona.prototype.getModelHierarchy = function (model) {
        var body = {
            personaId: this._personaId,
            model: model,
        };
        return this._scene.sendRequest('getModelHierarchy', body);
    };
    /**
     * Monitor bl variables.
     * @internal
     */
    Persona.prototype.createMonitorSet = function (setName, variables) {
        var body = {
            personaId: this._personaId,
            setName: [{ SetName: setName }],
            variables: variables,
        };
        return this._scene.sendRequest('createMonitorSet', body);
    };
    /**
     * @internal
     */
    Persona.prototype.removeMonitorSet = function (setName) {
        var body = {
            personaId: this._personaId,
            setName: [{ SetName: setName }],
        };
        return this._scene.sendRequest('removeMonitorSet', body);
    };
    /**
     * @internal
     */
    Persona.prototype.addVariableToMonitorSet = function (setName, variables) {
        var body = {
            personaId: this._personaId,
            setName: [{ SetName: setName }],
            variables: variables,
        };
        return this._scene.sendRequest('addVariableToMonitorSet', body);
    };
    /**
     * @internal
     */
    Persona.prototype.removeVariableFromMonitorSet = function (setName, variables) {
        var body = {
            personaId: this._personaId,
            setName: [{ SetName: setName }],
            variables: variables,
        };
        return this._scene.sendRequest('removeVariableFromMonitorSet', body);
    };
    /**
     * @internal
     */
    Persona.prototype.renderModel = function (modelName) {
        var body = {
            personaId: this._personaId,
            modelName: modelName,
        };
        return this._scene.sendRequest('renderModel', body);
    };
    Object.defineProperty(Persona.prototype, "onConversationResultEvent", {
        get: function () {
            return this._onConversationResultEvent;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Persona.prototype, "onSpeechMarkerEvent", {
        get: function () {
            return this._onSpeechMarkerEvent;
        },
        enumerable: false,
        configurable: true
    });
    return Persona;
}());
export { Persona };
