// @soulmachines/smwebsdk/lib-esm/Scene.js
/**
 * @module smwebsdk
 */
import { __assign, __awaiter, __generator } from "tslib";
/*
 * Copyright 2017-2020 Soul Machines Ltd. All Rights Reserved.
 */
import { Persona } from './Persona';
import { SmEvent } from './SmEvent';
import { LocalSession } from './LocalSession';
import { Session } from './Session';
import { WebSocketSession } from './WebSocketSession';
import { WebsocketCategory, WebsocketKind, } from './websocket-message/index';
import { SpeechState } from './websocket-message/enums/SpeechState';
import { SceneResponseError, } from './websocket-message/scene/index';
import { FeatureFlag } from './websocket-message/scene/response-body/StateResponseBody';
import { ContentAwareness } from './ContentAwareness';
import { Logger } from './utils/Logger';
import { ROOT_CONTEXT } from '@opentelemetry/api';
import { makeError } from './utils/make-error';
import convertToUserMedia from './utils/convertToUserMedia';
import { convertWssToHttps, getUrlHost } from './utils/utils';
import { UserMedia, } from './types/scene';
import { Conversation } from './Conversation';
import { MetadataSender } from './MetadataSender';
import { ConnectionState } from './ConnectionState';
import { SmTracerProvider } from './SmTelemetry';
import { websdkVersion } from './env-vars';
function sleep(t) {
    return new Promise(function (resolve) { return setTimeout(function () { return resolve(); }, t); });
}
var DEFAULT_RETRY_COUNT = 50;
var DEFAULT_RETRY_DELAY = 200;
var DEFAULT_PERSONA_ID = 1;
function retry(task, retryOptions, scene) {
    if (retryOptions === void 0) { retryOptions = {}; }
    return __awaiter(this, void 0, void 0, function () {
        var errors, count, delay, result, i, error_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    errors = [];
                    count = retryOptions.maxRetries || DEFAULT_RETRY_COUNT;
                    delay = retryOptions.delayMs || DEFAULT_RETRY_DELAY;
                    i = 0;
                    _a.label = 1;
                case 1:
                    if (!(i < count)) return [3 /*break*/, 8];
                    _a.label = 2;
                case 2:
                    _a.trys.push([2, 4, , 6]);
                    return [4 /*yield*/, task()];
                case 3:
                    result = _a.sent();
                    // store the result on the scene object
                    scene.connectionResult = {
                        message: 'success',
                        value: result,
                        retries: errors,
                    };
                    return [3 /*break*/, 6];
                case 4:
                    error_1 = _a.sent();
                    // collect a history of errors encountered during connect
                    errors.push(error_1);
                    // store the result on the scene object
                    scene.connectionResult = {
                        message: 'failed',
                        retries: errors,
                    };
                    //if error is 'noResumeSession` should cleanup session storage
                    if (error_1 instanceof Error && error_1.name === 'noSessionToResume') {
                        clearSessionData();
                    }
                    // any error other than 'noScene' should throw immediately
                    // and should not retry repeatedly.
                    // allows for proper errors and also string errors
                    if (!(error_1 instanceof Error) || error_1.name !== 'noScene') {
                        throw error_1;
                    }
                    // when we have reached the max number of retries,
                    // we should give up and throw the error
                    if (errors.length === count) {
                        console.warn("Retry gave up after ".concat(count, " retries:\n").concat(errors
                            .map(function (e) {
                            return e instanceof Error ? e.message : e.toString();
                        })
                            .join('\n')));
                        // throw the most recent error as the primary cause of failure
                        throw error_1;
                    }
                    return [4 /*yield*/, sleep(delay)];
                case 5:
                    _a.sent();
                    return [3 /*break*/, 7];
                case 6: return [3 /*break*/, 8];
                case 7:
                    i++;
                    return [3 /*break*/, 1];
                case 8: return [2 /*return*/, result];
            }
        });
    });
}
function storeSessionData(server, sessionId, apiKey) {
    sessionStorage.setItem('sm-server', server);
    sessionStorage.setItem('sm-session-id', sessionId);
    sessionStorage.setItem('sm-api-key', apiKey);
}
function getSessionData() {
    return {
        server: sessionStorage.getItem('sm-server'),
        resumeSessionId: sessionStorage.getItem('sm-session-id'),
        savedApiKey: sessionStorage.getItem('sm-api-key'),
    };
}
function clearSessionData() {
    sessionStorage.removeItem('sm-server');
    sessionStorage.removeItem('sm-session-id');
    sessionStorage.removeItem('sm-api-key');
}
/**
 * Scene class to hold a webrtc connection to a scene containing a persona.
 * @public
 */
var Scene = /** @class */ (function () {
    function Scene(videoOrOptions, audioOnly, requestedUserMedia, requiredUserMedia, contentAwarenessDebounceTime, loggingConfig, tracerOptions) {
        if (audioOnly === void 0) { audioOnly = false; }
        if (requestedUserMedia === void 0) { requestedUserMedia = UserMedia.MicrophoneAndCamera; }
        if (requiredUserMedia === void 0) { requiredUserMedia = UserMedia.Microphone; }
        var _this = this;
        this._onConversationResultEvents = {}; // persona id -> SmEvent with function(persona, result)
        this._onSpeechMarkerEvents = {}; // persona id -> SmEvent function(persona, marker)
        this._session = undefined;
        this._isWebSocketOnly = false;
        this._transactionId = 0;
        this._pendingResponses = {};
        this._microphoneUnmuteTimer = undefined;
        this._echoCancellationEnabled = true;
        this._serverControlledCameras = false;
        this._stopSpeakingWhenNotVisible = true;
        this._loggingConfig = {
            session: {},
            contentAwareness: {},
        };
        this._logger = new Logger();
        this._tracerOptions = {
            disableTracing: false,
            parentCtx: ROOT_CONTEXT,
            url: SmTracerProvider.defaultUrl,
        };
        this._sessionResumeEnabled = false;
        this._isResumedSession = false;
        this._sendMetadata = { pageUrl: false };
        this._onMicrophoneActive = new SmEvent();
        this._onCameraActive = new SmEvent();
        this.currentPersonaId = DEFAULT_PERSONA_ID;
        /** Returns the version of the webSdk and platformSdk */
        this.version = {
            webSdk: websdkVersion,
            platformSdk: 'unknown',
        };
        this.iosVisibilityChange = function () {
            var visible = document.visibilityState === 'visible';
            setTimeout(function () {
                if (_this._session) {
                    _this._session.sendRtcEvent('ui', { visible: visible });
                }
            }, 500); // allow 100ms for the H.264 decoder to become fully available again
        };
        this.stopSpeakingWhenNotVisible = function () {
            if (document.visibilityState !== 'visible') {
                _this.sendRequest('stopSpeaking', { personaId: _this.currentPersonaId });
            }
        };
        this.stopSpeakingWhenUnloaded = function () {
            _this.sendRequest('stopSpeaking', { personaId: _this.currentPersonaId });
        };
        // use the first parameter of the constructor to figure out
        // whether it was constructed using SceneOptions, or using
        // the deprecated multi-param format.
        if (this.isSceneOptions(videoOrOptions)) {
            // pull all private property initial values from the config object,
            // with fallbacks to the constructor property defaults if not provided
            var options = videoOrOptions;
            this._videoElement = options.videoElement;
            this._apiKey = options.apiKey;
            this._audioOnly = options.audioOnly || audioOnly;
            // default is "true" so can't use a shorthand falsy assessment to read this config option
            if (options.stopSpeakingWhenNotVisible === false) {
                this._stopSpeakingWhenNotVisible = false;
            }
            this._requestedUserMedia = convertToUserMedia(options.requestedMediaDevices, requestedUserMedia);
            this._requiredUserMedia = convertToUserMedia(options.requiredMediaDevices, requiredUserMedia);
            this.contentAwarenessDebounceTime = options.contentAwarenessDebounceTime;
            this._loggingConfig = __assign(__assign({}, this._loggingConfig), (options.loggingConfig || {}));
            if (options.sendMetadata) {
                this._sendMetadata = options.sendMetadata;
            }
            if (options.tracerOptions) {
                this._tracerOptions = options.tracerOptions;
            }
        }
        else {
            // take all private property initial values directly from the constructor props
            this._videoElement = videoOrOptions;
            this._audioOnly = audioOnly;
            this._requestedUserMedia = requestedUserMedia;
            this._requiredUserMedia = requiredUserMedia;
            this.contentAwarenessDebounceTime = contentAwarenessDebounceTime;
            this._loggingConfig = __assign(__assign({}, this._loggingConfig), loggingConfig);
            if (tracerOptions) {
                this._tracerOptions = tracerOptions;
            }
        }
        this._logger = new Logger(this._loggingConfig.session.minLogLevel, this._loggingConfig.session.enabled);
        /**
         * call onStateEvent.addListener(function(scene, state)) to be called when a state message is received as per the scene protocol
         * call onStateEvent.removeListener(function(scene, state)) to deregister a listener.
         */
        this._onStateEvent = new SmEvent();
        this._onStateEvent.addListener(function (scene, state) {
            if (_this._onState) {
                _this._onState(scene, state);
            }
        });
        /**
         * call onRecognizeResultsEvent.addListener(function(scene, status, errorMessage, results)) to be called when speech to text results are recognized, results are documented in scene protocol.
         * call onRecognizeResultsEvent.removeListener(function(scene, status, errorMessage, results)) to deregister a listener.
         */
        this._onRecognizeResultsEvent = new SmEvent();
        this._onRecognizeResultsEvent.addListener(function (scene, status, errorMessage, results) {
            if (_this._onRecognizeResults) {
                _this._onRecognizeResults(scene, status, errorMessage, results);
            }
        });
        /**
         * call onDisconnectedEvent.addListener(function(scene, sessionId, reason)) to be  called when the session is disconnected.
         * call onDisconnectedEvent.removeListener(function(scene, sessionId, reason)) to deregister a listener.
         */
        this._onDisconnectedEvent = new SmEvent();
        this._onDisconnectedEvent.addListener(function (scene, sessionId, reason) {
            clearSessionData();
            _this.cleanupEventListeners();
            if (_this._onDisconnected) {
                _this._onDisconnected(scene, sessionId, reason);
            }
        });
        /**
         * call onUserTextEvent.addListener(function(scene, text)) to be called when a custom text message is sent from the orchestration server
         * call onUserTextEvent.removeListener(function(scene, text)) to deregister a listener.
         */
        this._onUserTextEvent = new SmEvent();
        this._onUserTextEvent.addListener(function (scene, text) {
            if (_this._onUserText) {
                _this._onUserText(scene, text);
            }
        });
        this._onDemoModeEvent = new SmEvent();
        this._underRuntimeHost = Boolean(window.SmIsUnderRuntimeHost);
        // Generate a random id for the scene. This is used internally with the _transactionId
        // to ensure unique transaction ids when mulitple Scene instances access the same BL instances
        // eg multiple Soul Studio windows
        var randArray = new Uint32Array(3);
        window.crypto.getRandomValues(randArray);
        this._sceneId = randArray.toString().replace(/,/g, '-');
        this.conversation = new Conversation();
        this.connectionState = new ConnectionState();
        this._metadataSender = new MetadataSender(this);
        this._logger.log('debug', 'websdk version:', this.version.webSdk);
    }
    /**
     * Tests the first value of the Scene construtor to decide if
     * it matches the new-style config options format.
     */
    Scene.prototype.isSceneOptions = function (videoOrOptions) {
        // scene options must be defined, even if they're an empty object
        var isDefined = !!videoOrOptions;
        // scene options object will not have a tagName
        var isHTMLElement = !!(videoOrOptions === null || videoOrOptions === void 0 ? void 0 : videoOrOptions.tagName);
        return isDefined && !isHTMLElement;
    };
    Scene.prototype.connectionValid = function () {
        if (this._underRuntimeHost) {
            return true;
        }
        if (this._session && this._session.serverConnection) {
            return true;
        }
        return false;
    };
    /**
     * Check if the scene connection is open and valid.
     *
     * @returns Returns true if the connection is open and valid otherwise false.
     */
    Scene.prototype.isConnected = function () {
        if (this.connectionValid() &&
            this._session &&
            this._session.serverConnection &&
            this._session.serverConnection.readyState ===
                this._session.serverConnection.OPEN) {
            return true;
        }
        return false;
    };
    /**
     * Extends the server side timeout. This also happens automatically whenever the persona speaks.
     */
    Scene.prototype.keepAlive = function () {
        if (this._session && this._session.peerConnection !== null) {
            this._session.sendRtcEvent('keepAlive', {});
        }
    };
    /**
     * Disconnects the session
     */
    Scene.prototype.disconnect = function () {
        var _a, _b;
        clearSessionData();
        this.cleanupEventListeners();
        this.connectionState.reset();
        this.conversation.reset();
        (_a = this.contentAwareness) === null || _a === void 0 ? void 0 : _a.disconnect();
        this._metadataSender.disconnect();
        (_b = this._session) === null || _b === void 0 ? void 0 : _b.close(true);
        this._session = undefined;
    };
    Scene.prototype.connect = function (serverUriOrOptions, userText, accessToken, retryOptions) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function () {
            var connectStartTime, config, span, recordEventsToSpan, hasPassedInTokenServerAuth, response, data, server, error_2, _d, initTracerStartTime, initTracerEndTime;
            var _this = this;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        connectStartTime = Date.now();
                        config = this.connectArgsToConfig(serverUriOrOptions, userText, accessToken, retryOptions);
                        recordEventsToSpan = function (event) {
                            span === null || span === void 0 ? void 0 : span.addEvent(event.name);
                        };
                        this.connectionState.onConnectionStateUpdated.addListener(recordEventsToSpan);
                        if (!this._underRuntimeHost) return [3 /*break*/, 1];
                        this._session = new LocalSession(this._videoElement, this._logger);
                        return [3 /*break*/, 8];
                    case 1:
                        hasPassedInTokenServerAuth = config.tokenServerUri || config.tokenServerAccessToken;
                        if (this._apiKey && hasPassedInTokenServerAuth) {
                            this._logger.log('warn', 'You are trying to connect via an API key and a token server. Please use one or the other');
                        }
                        if (!(this._apiKey && !hasPassedInTokenServerAuth)) return [3 /*break*/, 6];
                        _e.label = 2;
                    case 2:
                        _e.trys.push([2, 5, , 6]);
                        return [4 /*yield*/, this.fetchAuthConfig(this._apiKey)];
                    case 3:
                        response = _e.sent();
                        return [4 /*yield*/, response.json()];
                    case 4:
                        data = _e.sent();
                        server = getSessionData().server;
                        config.tokenServerUri = data.url;
                        config.tokenServerAccessToken = data.jwt;
                        if (server) {
                            config.tokenServerUri = getUrlHost(data.url) + 'server/' + server;
                        }
                        return [3 /*break*/, 6];
                    case 5:
                        error_2 = _e.sent();
                        if (error_2 instanceof Error && error_2.message === 'Broken API key') {
                            this._logger.log('error', 'Broken API key. Please check your key or re copy the key from DDNA Studio.');
                        }
                        else {
                            this._logger.log('error', 'Invalid API key: Please check your key configuration in DDNA Studio. For more information click here https://soulmachines-support.atlassian.net/wiki/spaces/SSAS/pages/1320058919/Connecting+Using+API+Keys#Troubleshooting');
                        }
                        throw makeError('Invalid API key', 'serverConnectionFailed');
                    case 6:
                        if (!config.tokenServerUri || !config.tokenServerAccessToken) {
                            throw makeError('Please authenticate via an API key or with a serverUri and accessToken', 'serverConnectionFailed');
                        }
                        return [4 /*yield*/, this.initializeTracer(config.tokenServerUri, config.tokenServerAccessToken)];
                    case 7:
                        _d = _e.sent(), initTracerStartTime = _d.initTracerStartTime, initTracerEndTime = _d.initTracerEndTime;
                        span = (_c = (_b = (_a = SmTracerProvider.getTracer()) === null || _a === void 0 ? void 0 : _a.startSpan('createSessionAndConnect')) === null || _b === void 0 ? void 0 : _b.setAttribute('sm.websdk.connection.pretraceinitduration.milliseconds', initTracerStartTime - connectStartTime)) === null || _c === void 0 ? void 0 : _c.setAttribute('sm.websdk.connection.traceinitduration.milliseconds', initTracerEndTime - initTracerStartTime);
                        if (this._isWebSocketOnly) {
                            this._session = new WebSocketSession(config.tokenServerUri, config.tokenServerAccessToken, this._logger);
                        }
                        else {
                            this._session = new Session(this._videoElement, config.tokenServerUri, config.userText, config.tokenServerAccessToken, this._audioOnly, this._requestedUserMedia, this._requiredUserMedia, this._echoCancellationEnabled, this._logger, this.connectionState);
                        }
                        _e.label = 8;
                    case 8:
                        if (!this._session) {
                            throw makeError('Failed to create session', 'unknown');
                        }
                        this._session.onConnected = this.sessionConnected.bind(this);
                        this._session.onMessage = this.onMessage.bind(this);
                        this._session.onClose = this.sessionClosed.bind(this);
                        this._session.onUserText = this.rtcUserText.bind(this);
                        if ('microphoneActiveCallbacks' in this._session) {
                            this._session.microphoneActiveCallbacks = this._onMicrophoneActive;
                        }
                        if ('cameraActiveCallbacks' in this._session) {
                            this._session.cameraActiveCallbacks = this._onCameraActive;
                        }
                        if (this._session.features.isIos) {
                            document.addEventListener('visibilitychange', this.iosVisibilityChange);
                        }
                        return [4 /*yield*/, retry(function () { return __awaiter(_this, void 0, void 0, function () {
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0: return [4 /*yield*/, this._session.connect()];
                                        case 1: 
                                        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                                        return [2 /*return*/, _a.sent()];
                                    }
                                });
                            }); }, config.retryOptions, this).finally(function () {
                                _this.connectionState.onConnectionStateUpdated.removeListener(recordEventsToSpan);
                                span === null || span === void 0 ? void 0 : span.end();
                            })];
                    case 9: return [2 /*return*/, _e.sent()];
                }
            });
        });
    };
    Scene.prototype.initializeTracer = function (tokenServerUri, tokenServerAccessToken) {
        return __awaiter(this, void 0, void 0, function () {
            var initTracerStartTime, error_3, base;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        initTracerStartTime = Date.now();
                        if (!(!this._tracerOptions.disableTracing &&
                            !SmTracerProvider.isInitialized())) return [3 /*break*/, 4];
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.initTelemetryToken({
                                tokenServerUri: tokenServerUri,
                                authToken: tokenServerAccessToken,
                            })];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        error_3 = _a.sent();
                        base = 'Could not initialize tracer telemetry token: ';
                        if (error_3 instanceof ReferenceError &&
                            error_3.message === 'fetch is not defined') {
                            // This happens in a non-browser environment (usually a test
                            // environment), so we can't get a telemetry token. Just log to
                            // debug.
                            this._logger.log('debug', base + error_3.name + ': ' + error_3.message);
                        }
                        else if (error_3 instanceof Error) {
                            this._logger.log('warn', base + error_3.name + ': ' + error_3.message);
                        }
                        else {
                            this._logger.log('warn', base + 'unknown error type');
                        }
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, { initTracerStartTime: initTracerStartTime, initTracerEndTime: Date.now() }];
                }
            });
        });
    };
    Scene.prototype.initTelemetryToken = function (_a) {
        var tokenServerUri = _a.tokenServerUri, authToken = _a.authToken;
        return __awaiter(this, void 0, void 0, function () {
            var otelJwtEndpoint, tracesEndpoint, telemetryHost, resp, otelResponse;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        otelJwtEndpoint = 'api/telemetry/jwt';
                        tracesEndpoint = 'api/telemetry/v1/traces';
                        telemetryHost = getUrlHost(convertWssToHttps(tokenServerUri));
                        if (!telemetryHost) {
                            this._logger.log('debug', 'Could not initialize tracer telemetry token: invalid token server URI');
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, fetch(telemetryHost + otelJwtEndpoint, {
                                headers: {
                                    Authorization: 'Bearer ' + authToken,
                                },
                            })];
                    case 1:
                        resp = _b.sent();
                        if (!resp) {
                            this._logger.log('warn', 'Failed to receive response from otel token endpoint');
                            return [2 /*return*/];
                        }
                        if (!resp.ok) {
                            this._logger.log('warn', 'Failed to fetch otel token: ' + resp.status + ': ' + resp.statusText);
                            return [2 /*return*/];
                        }
                        return [4 /*yield*/, resp.json()];
                    case 2:
                        otelResponse = _b.sent();
                        if (!otelResponse.success) {
                            this._logger.log('warn', 'Failed: otel response not successful');
                            return [2 /*return*/];
                        }
                        try {
                            SmTracerProvider.init({
                                jwt: otelResponse.telemetryJwt,
                                url: telemetryHost + tracesEndpoint,
                                webSDKVersion: this.version.webSdk,
                            });
                            this._logger.log('log', 'Telemetry initialized');
                        }
                        catch (error) {
                            if (error instanceof Error) {
                                this._logger.log('warn', 'Failed to initialize tracer: ' + error.message);
                            }
                            else {
                                this._logger.log('warn', 'Failed to initialize tracer: unknown error');
                            }
                        }
                        return [2 /*return*/];
                }
            });
        });
    };
    Scene.prototype.onMessage = function (message) {
        var category = message.category;
        if (category === 'scene') {
            var sceneMessage = message;
            this.onSceneMessage(sceneMessage);
            return;
        }
    };
    Scene.prototype.sendOnewaySceneRequest = function (name, body) {
        if (!this._session) {
            return;
        }
        var payload = {
            name: name,
            body: body,
            category: WebsocketCategory.Scene,
            kind: WebsocketKind.Request,
        };
        this._session.sendMessage(payload);
    };
    /**
     * The internal method used for sending request messages.
     *
     * All offically supported message have their own public methods (e.g. `conversationSend()` or `scene.startRecognize()`). \
     * Please use those instead.
     *
     * @internal
     */
    Scene.prototype.sendRequest = function (name, body) {
        var _this = this;
        if (body === void 0) { body = {}; }
        return new Promise(function (resolve, reject) {
            if (!_this._session) {
                reject(makeError('No session available', 'noSession'));
                return;
            }
            var transaction = _this._sceneId + '_' + ++_this._transactionId;
            var payload = {
                transaction: transaction,
                name: name,
                body: body,
                category: WebsocketCategory.Scene,
                kind: WebsocketKind.Request,
            };
            var pending = { resolve: resolve, reject: reject };
            _this._pendingResponses[transaction] = pending;
            if (_this._session) {
                _this._session.sendMessage(payload);
            }
        });
    };
    Scene.prototype.onSceneMessage = function (message) {
        var _a, _b, _c;
        var name = message.name, body = message.body, kind = message.kind, status = message.status, transaction = message.transaction;
        // Process events
        if (body && name === 'state') {
            var responseBody = body;
            this._onStateEvent.call(this, responseBody);
            if ((_a = responseBody.scene) === null || _a === void 0 ? void 0 : _a.featureFlags) {
                this.enableFlaggedFeatures(responseBody.scene.featureFlags);
            }
            if ((_b = responseBody.scene) === null || _b === void 0 ? void 0 : _b.sdkVersion) {
                this.version.platformSdk = (_c = responseBody.scene) === null || _c === void 0 ? void 0 : _c.sdkVersion;
                this._logger.log('debug', 'platformSdk version:', this.version.platformSdk);
            }
            this.conversation.processStateMessage(responseBody);
            // mute the microphone while a persona is speaking
            this.controlMicrophoneMute(body);
        }
        else if (body && name === 'recognizeResults') {
            var _d = body, status_1 = _d.status, errorMessage = _d.errorMessage, results = _d.results;
            this.conversation.processRecognizeResultsMessage(body);
            this._onRecognizeResultsEvent.call(this, status_1, errorMessage, results);
        }
        else if (body && name === 'conversationResult') {
            this.conversation.onConversationResult(body);
            var personaId = body.personaId;
            if (personaId) {
                var persona = new Persona(this, personaId);
                var event_1 = this._onConversationResultEvents[personaId];
                event_1.call(persona, body);
                this.currentPersonaId = personaId;
            }
        }
        else if (body && name === 'speechMarker') {
            this.conversation.onSpeechMarker(body);
            var personaId = body.personaId;
            if (personaId) {
                var persona = new Persona(this, personaId);
                var event_2 = this._onSpeechMarkerEvents[personaId];
                event_2.call(persona, body);
                this.currentPersonaId = personaId;
            }
        }
        else if (body && name === 'demoMode') {
            this._onDemoModeEvent.call(this, body);
        }
        // Process responses, message should always be a response as far as we're aware
        if (kind === WebsocketKind.Response && transaction) {
            this.processResponse(body, name, status, transaction);
        }
    };
    Scene.prototype.processResponse = function (body, name, status, transaction) {
        // Check for a pending response
        var pending = this._pendingResponses[transaction];
        if (pending) {
            if (status === 0) {
                // Success
                pending.resolve(body);
            }
            else {
                // Failure
                var error = new SceneResponseError();
                error.requestName = name;
                error.status = status;
                error.responseBody = body;
                pending.reject(error);
            }
            delete this._pendingResponses[transaction];
        }
    };
    Scene.prototype.controlMicrophoneMute = function (state) {
        var _this = this;
        // Watch for speaking state transitions and mute the
        // microphone during persona speech to prevent self interruption
        if (state.persona &&
            this._session &&
            this._session.microphoneMuteDelay !== -1) {
            // iterate through the personas looking for speaking state changes
            for (var personaId in state.persona) {
                var persona_state = state.persona[personaId];
                if (!persona_state.speechState) {
                    continue;
                }
                if (persona_state.speechState === SpeechState.Speaking) {
                    // A persona is speaking, mute the microphone
                    if (!this._session.microphoneMuted) {
                        this._logger.log('warn', 'Persona is speaking - mute microphone');
                        this._session.microphoneMuted = true;
                    }
                    if (this._microphoneUnmuteTimer) {
                        // ensure an in-progress timeout doesn't incorrectly unmute
                        clearTimeout(this._microphoneUnmuteTimer);
                        this._microphoneUnmuteTimer = undefined;
                    }
                }
                else {
                    // A persona has stopped speaking, unmute the microphone after
                    // the microphone mute delay
                    if (this._session.microphoneMuted && !this._microphoneUnmuteTimer) {
                        this._microphoneUnmuteTimer = setTimeout(function () {
                            if (!_this._session || !_this._microphoneUnmuteTimer) {
                                return;
                            }
                            _this._logger.log("warn", 'Persona is no longer speaking - unmute microphone');
                            _this._session.microphoneMuted = false;
                            _this._microphoneUnmuteTimer = undefined;
                        }, this._session.microphoneMuteDelay);
                    }
                }
            }
        }
    };
    /** Close the current scene connection */
    Scene.prototype.close = function () {
        // close/disconnect the session
        if (this._session) {
            this._session.close(true);
        }
    };
    Scene.prototype.sessionConnected = function (resumeRequested, isResumedSession, server, sessionId) {
        this.contentAwareness = new ContentAwareness(this, this.contentAwarenessDebounceTime, new Logger(this._loggingConfig.contentAwareness.minLogLevel, this._loggingConfig.contentAwareness.enabled));
        /*
        Interrupt DP from speaking so it does not continue talking when user switches tabs
        */
        if (this._stopSpeakingWhenNotVisible) {
            document.addEventListener('visibilitychange', this.stopSpeakingWhenNotVisible);
        }
        //when user navigates to a new page, widget DP should stop speaking to allow new welcome message coming through in the new page
        //however browser doesn't trigger visibilitychange event, only "beforeunload" event is sure to be triggered
        window.addEventListener('beforeunload', this.stopSpeakingWhenUnloaded);
        if (this._sendMetadata.pageUrl) {
            this._metadataSender.observeUrlChanges();
        }
        //update resume session data
        this._isResumedSession = isResumedSession;
        if (resumeRequested) {
            this._sessionResumeEnabled = true;
            storeSessionData(server, sessionId, this._apiKey || '');
        }
        // When page navigation happens, check if any value in _sendMetadata is true and send it back to NLP so conversation writers can use it
        if (isResumedSession && this._sendMetadata.pageUrl) {
            this._metadataSender.send();
        }
    };
    Scene.prototype.cleanupEventListeners = function () {
        var _a;
        if ((_a = this._session) === null || _a === void 0 ? void 0 : _a.features.isIos) {
            document.removeEventListener('visibilitychange', this.iosVisibilityChange);
        }
        if (this._stopSpeakingWhenNotVisible) {
            document.removeEventListener('visibilitychange', this.stopSpeakingWhenNotVisible);
        }
        window.removeEventListener('beforeunload', this.stopSpeakingWhenUnloaded);
    };
    Scene.prototype.sessionClosed = function (reason) {
        clearSessionData();
        this.cleanupEventListeners();
        if (this._session) {
            this.connectionState.reset();
            this.conversation.reset();
            this._onDisconnectedEvent.call(this, this._session.sessionId, reason);
        }
    };
    Scene.prototype.rtcUserText = function (text) {
        this._onUserTextEvent.call(this, text);
    };
    Scene.prototype.enableFlaggedFeatures = function (featureFlags) {
        this._serverControlledCameras = featureFlags.includes(FeatureFlag.UI_SDK_CAMERA_CONTROL);
    };
    Scene.prototype.sendContent = function () {
        var _a;
        if (!this.contentAwareness) {
            console.warn('ContentAwareness is not enabled for this project');
        }
        (_a = this.contentAwareness) === null || _a === void 0 ? void 0 : _a.measure();
    };
    /**
     * Sends updated video element size to server
     * this gives the app the chance to choose what size should be rendered on server
     * and the application is responsible to register for a video element size change
     * event and call this method to maintain best possible video quality for the size
     * and/or to set an updated video element size and then call this method.
     * @param width - The width in pixels to render the video
     * @param height - The height in pixels to render the video
     */
    Scene.prototype.sendVideoBounds = function (width, height) {
        if (this._session) {
            this._session.sendVideoBounds(width, height);
        }
    };
    /**
     * Send configuration to the scene
     * @param configuration - Scene configuration as per the scene protocol
     */
    Scene.prototype.configure = function (configuration) {
        return this.sendRequest('configure', configuration);
    };
    /**
     * Send a custom user text message to the orchestration server
     * @param text - Custom text sent to the orchestration server
     */
    Scene.prototype.sendUserText = function (text) {
        if (this._session) {
            this._session.sendUserText(text);
        }
    };
    /**
     * Start the speech to text recognizer
     * @param audioSource - The audio source either smwebsdk.audioSource.processed or
     *                    smwebsdk.audioSource.squelched, defaults to processed.
     */
    Scene.prototype.startRecognize = function (audioSource) {
        var body = {};
        if (audioSource !== undefined) {
            body.audioSource = audioSource;
        }
        return this.sendRequest('startRecognize', body);
    };
    /** Stop the speech to text reconizer */
    Scene.prototype.stopRecognize = function () {
        return this.sendRequest('stopRecognize');
    };
    /** Is the microphone connected in the session */
    Scene.prototype.isMicrophoneConnected = function () {
        // public function rather than getter for back compatibility
        if (this._session) {
            return this._session.isMicrophoneConnected;
        }
        return false;
    };
    /** Is the camera connected in the session */
    Scene.prototype.isCameraConnected = function () {
        // public function rather than getter for back compatibility
        if (this._session) {
            return this._session.isCameraConnected;
        }
        return false;
    };
    Scene.prototype.session = function () {
        // public function rather than getter for back compatibility
        return this._session;
    };
    Scene.prototype.hasContentAwareness = function () {
        return !!this.contentAwareness;
    };
    Scene.prototype.hasServerControlledCameras = function () {
        return this._serverControlledCameras;
    };
    /**
     * Check if session persistence feature is supported in current session
     *
     * @returns `boolean`
     *
     * Usage:
     * ```javascript
     * const isSessionPersistenceSupported = scene.supportsSessionPersistence();
     * ```
     */
    Scene.prototype.supportsSessionPersistence = function () {
        return this._sessionResumeEnabled;
    };
    /**
     * Check if current session is a new session or a resumed session
     *
     * @returns `boolean`
     *
     * Usage:
     * ```javascript
     * const isResumedSession = scene.isResumedSession();
     * ```
     */
    Scene.prototype.isResumedSession = function () {
        return this._isResumedSession;
    };
    Object.defineProperty(Scene.prototype, "onConversationResultEvents", {
        get: function () {
            return this._onConversationResultEvents;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "onSpeechMarkerEvents", {
        get: function () {
            return this._onSpeechMarkerEvents;
        },
        enumerable: false,
        configurable: true
    });
    /** Get the current scene state */
    Scene.prototype.getState = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.sendRequest('getState')];
            });
        });
    };
    Object.defineProperty(Scene.prototype, "onStateEvent", {
        get: function () {
            return this._onStateEvent;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "onState", {
        // eslint-disable-next-line @typescript-eslint/ban-types
        set: function (onState) {
            this._onState = onState;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "onDisconnectedEvent", {
        get: function () {
            return this._onDisconnectedEvent;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "onDisconnected", {
        // eslint-disable-next-line @typescript-eslint/ban-types
        set: function (onDisconnected) {
            this._onDisconnected = onDisconnected;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "onRecognizeResultsEvent", {
        get: function () {
            return this._onRecognizeResultsEvent;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "onRecognizeResults", {
        // eslint-disable-next-line @typescript-eslint/ban-types
        set: function (onRecognizeResults) {
            this._onRecognizeResults = onRecognizeResults;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "onUserTextEvent", {
        get: function () {
            return this._onUserTextEvent;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "onUserText", {
        // eslint-disable-next-line @typescript-eslint/ban-types
        set: function (onUserText) {
            this._onUserText = onUserText;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "echoCancellationEnabled", {
        get: function () {
            return this._echoCancellationEnabled;
        },
        set: function (enabled) {
            this._echoCancellationEnabled = enabled;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "onDemoModeEvent", {
        /**
         * @internal
         */
        get: function () {
            return this._onDemoModeEvent;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "videoElement", {
        get: function () {
            return this._videoElement;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "viewerOffsetX", {
        get: function () {
            if (this._session) {
                return this._session.offsetX;
            }
            return 0;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "viewerOffsetY", {
        get: function () {
            if (this._session) {
                return this._session.offsetY;
            }
            return 0;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "isWebSocketOnly", {
        get: function () {
            return this._isWebSocketOnly;
        },
        set: function (isWebSocketOnly) {
            this._isWebSocketOnly = isWebSocketOnly;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Scene.prototype, "onMicrophoneActive", {
        /**
         * @returns an {@link SmEvent} associated with the microphone.
         *
         * Listeners can then be added to this event allowing you to call functions when the microphone active status changes.
         *
         * Usage:
         * ```javascript
         * scene.onMicrophoneActive.addListener(
         *   (active) => console.log('Microphone Active: ', active));
         * ```
         */
        get: function () {
            return this._onMicrophoneActive;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Specifies if the microphone is currently active and streaming audio
     * to the server.
     *
     * @returns `boolean`
     *
     * Usage:
     * ```javascript
     * const isMicrophoneActive = scene.isMicrophoneActive();
     * ```
     */
    Scene.prototype.isMicrophoneActive = function () {
        var _a;
        return Boolean((_a = this._session) === null || _a === void 0 ? void 0 : _a.isMicrophoneActive());
    };
    Object.defineProperty(Scene.prototype, "onCameraActive", {
        /**
         * @returns an {@link SmEvent} associated with the camera.
         *
         * Listeners can then be added to this event allowing you to call functions when the camera active status changes.
         *
         * Usage:
         * ```javascript
         * scene.onCameraActive.addListener(
         *   (active) => console.log('Camera Active: ', active));
         * ```
         */
        get: function () {
            return this._onCameraActive;
        },
        enumerable: false,
        configurable: true
    });
    /**
     * Specifies if the camera is currently active and streaming video
     * to the server.
     *
     * @returns `boolean`
     *
     * Usage:
     * ```javascript
     * const isCameraActive = scene.isCameraActive();
     * ```
     */
    Scene.prototype.isCameraActive = function () {
        var _a;
        return Boolean((_a = this._session) === null || _a === void 0 ? void 0 : _a.isCameraActive());
    };
    /**
     * On success, starts or stops streaming video/audio to the server based on the values of `microphone` and `camera`.
     *
     * @param options.microphone - If `true`, activates the microphone and starts streaming audio. \
     * If `false` deactivates the microphone and stops streaming audio. \
     * If not set, microphone will retain its existing state.
     * @param options.camera - If `true`, activates the camera and starts streaming video. \
     * If `false` deactivates the camera and stops streaming video. \
     * If not set, microphone will retain its existing state.
     *
     * @returns Returns a promise which is fulfilled when the media active state has been successfully changed. \
     * If the session is not defined it will return `undefined`. \
     * If the active state could not be changed, the promise is rejected with an Error object having the format:
     * ```javascript
     * {
     *   message: string;
     *   name: errorCode;
     * }
     * ```
     * Where `errorCode` is one of:
     *    - `noUserMedia` - the microphone/camera is either not available, not usable or the user declined permission to use them
     *    - `failedUpgrade` - the media upgrade failed
     *    - `notSupported` - user’s browser does not support the getUserMedia API
     *    - `noConnection` - connection has not been established - ensure scene.connect() has been called previously
     *
     * Usage:
     * ```javascript
     * scene.setMediaDeviceActive({ microphone: true, camera: false })
     *   .then(console.log('microphone activated, camera deactivated'));
     *   .catch((error) => console.log('error occurred: ', error);
     * ```
     */
    Scene.prototype.setMediaDeviceActive = function (options) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.isConnected()) return [3 /*break*/, 2];
                        return [4 /*yield*/, ((_a = this._session) === null || _a === void 0 ? void 0 : _a.setMediaDeviceActive({
                                microphone: options.microphone,
                                camera: options.camera,
                            }))];
                    case 1:
                        _b.sent();
                        return [3 /*break*/, 3];
                    case 2: throw makeError('Connection has not been established', 'noConnection');
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Play the video element and return results. Different browsers have different restrictions on autoplay.
     * Using this method can handle all the cases browsers can have on inital video playback.
     * @param videoElement - Optional parameter specifying the video element hosting the Digital Person. If not specified the video element passed to the Scene constructor will be used.
     * @returns Returns a promise which is fulfilled when the video playback is successful, with indication of video and audio status.
     * If the video element is not defined or video play fails the promise is rejected with an Error object having the format:
     * ```javascript
     * {
     *   message: string;
     *   name: errorCode;
     * }
     * ```
     * Where `errorCode` is one of:
     *    - `noVideoElement` - no HTMLVideoElement found from `videoElement` or `Scene` constructor
     *    - `userInteractionRequired` - cannot start media playback due to browser restriction; user interaction is required before playing again
     *
     * Usage:
     * ```javascript
     * scene.startVideo()
     *      .then(({ video, audio }) => {
     *         if (!audio) {
     *          //video is muted, ask user to unmute video
     *         }
     *      })
     *      .catch((error) => {
     *         if (error.name === 'userInteractionRequired') {
     *          //ask user to interact with the UI
     *          //unmute video and play again
     *          video.muted = false;
     *          video.play();
     *         }
     *      });
     * ```
     */
    Scene.prototype.startVideo = function (videoElement) {
        return __awaiter(this, void 0, void 0, function () {
            var video;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        video = videoElement || this._videoElement;
                        if (!video) {
                            throw makeError('Cannot find HTMLVideoElement', 'noVideoElement');
                        }
                        return [4 /*yield*/, this.playVideo(video)];
                    case 1:
                        // best case, play with audio
                        if (_a.sent()) {
                            return [2 /*return*/, {
                                    video: true,
                                    audio: true,
                                }];
                        }
                        //second-best case, play without audio
                        video.muted = true;
                        return [4 /*yield*/, this.playVideo(video)];
                    case 2:
                        if (_a.sent()) {
                            return [2 /*return*/, {
                                    video: true,
                                    audio: false,
                                }];
                        }
                        //worst case, not able to play, require user interaction
                        throw makeError('Cannot start media playback', 'userInteractionRequired');
                }
            });
        });
    };
    Scene.prototype.playVideo = function (videoElement) {
        return __awaiter(this, void 0, void 0, function () {
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, videoElement.play()];
                    case 1:
                        _b.sent();
                        return [2 /*return*/, true];
                    case 2:
                        _a = _b.sent();
                        return [2 /*return*/, false];
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    Scene.prototype.fetchAuthConfig = function (apiKey) {
        var authServer;
        try {
            var decodedApiKey = JSON.parse(atob(apiKey));
            authServer = decodedApiKey.authServer;
        }
        catch (error) {
            throw makeError('Broken API key', 'Failed to decode api key');
        }
        // check if sessionId exists in browser storage
        var _a = getSessionData(), server = _a.server, resumeSessionId = _a.resumeSessionId, savedApiKey = _a.savedApiKey;
        // check if the current api key is same as saved api key (if it is intended to resume to the same DP)
        if (server && resumeSessionId && savedApiKey === apiKey) {
            authServer = authServer + '?sessionId=' + resumeSessionId;
        }
        return fetch(authServer, {
            headers: {
                key: apiKey,
            },
        });
    };
    // Use the first parameter of the constructor to figure out
    // whether it was constructed using ConnectOptions, or using the deprecated multi-param format.
    Scene.prototype.connectArgsToConfig = function (serverUriOrOptions, userText, accessToken, retryOptions) {
        var _a, _b;
        if (typeof serverUriOrOptions === 'string') {
            return {
                tokenServerUri: serverUriOrOptions,
                tokenServerAccessToken: accessToken,
                userText: userText,
                retryOptions: retryOptions,
            };
        }
        else {
            return {
                tokenServerUri: ((_a = serverUriOrOptions === null || serverUriOrOptions === void 0 ? void 0 : serverUriOrOptions.tokenServer) === null || _a === void 0 ? void 0 : _a.uri) || '',
                tokenServerAccessToken: (_b = serverUriOrOptions === null || serverUriOrOptions === void 0 ? void 0 : serverUriOrOptions.tokenServer) === null || _b === void 0 ? void 0 : _b.token,
                userText: serverUriOrOptions === null || serverUriOrOptions === void 0 ? void 0 : serverUriOrOptions.userText,
                retryOptions: serverUriOrOptions === null || serverUriOrOptions === void 0 ? void 0 : serverUriOrOptions.retryOptions,
            };
        }
    };
    /**
     * Check if the session logging is enabled.
     *
     * @returns Returns true if the session logging is enabled otherwise false.
     */
    Scene.prototype.isLoggingEnabled = function () {
        return this._logger.isEnabled;
    };
    /**
     * Check minimal log level of session logging.
     *
     * @returns Returns minimal log setting of session logging, type is LogLevel.
     */
    Scene.prototype.getMinLogLevel = function () {
        return this._logger.getMinLogLevel();
    };
    /**
     * Enable/disable session logging
     * @param enable - set true to enable session log, false to disable
     */
    Scene.prototype.setLogging = function (enable) {
        this._logger.enableLogging(enable);
    };
    /**
     * Set minimal log level of session logging.
     * @param level - use LogLevel type to set minimal log level of session logging
     */
    Scene.prototype.setMinLogLevel = function (level) {
        this._logger.setMinLogLevel(level);
    };
    return Scene;
}());
export { Scene };
