// @soulmachines/smwebsdk/lib-esm/WebSocketSession.js
/**
 * @module smwebsdk
 */
import { __awaiter, __generator } from "tslib";
/*
 * Copyright 2020 Soul Machines Ltd. All Rights Reserved.
 */
import { Deferred } from './Deferred';
import { Features } from './Features';
import { Logger } from './utils/Logger';
import { makeError } from './utils/make-error';
/**
 *  WebSocketSession class
 */
var WebSocketSession = /** @class */ (function () {
    function WebSocketSession(serverUri, accessToken, logger) {
        if (logger === void 0) { logger = new Logger(); }
        var _this = this;
        this.logger = logger;
        this._outgoingQueue = [];
        this._onConnectedStorage = function (resumeRequested, isResumedSession, server, sessionId
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        ) { };
        this._pendingLog = [];
        this._closed = false;
        this._shouldLogToServer = false;
        this._serverUri = serverUri;
        this._accessToken = accessToken;
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        this._onClose = function (reason) { }; // owner specifies custom close method
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        this._onMessage = function (message) { }; // owner specifies custom message handler
        this._sessionError = function (error) {
            // owner can specify custom session error handler
            _this.logger.log('error', "session error: ".concat(error));
        };
        this._features = new Features();
    }
    Object.defineProperty(WebSocketSession.prototype, "onConnected", {
        set: function (sessionFunction) {
            this._onConnectedStorage = sessionFunction;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "onClose", {
        set: function (closeFunction) {
            this._onClose = closeFunction;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "onMessage", {
        set: function (messageFunction) {
            this._onMessage = messageFunction;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "loggingEnabled", {
        get: function () {
            return this.logger.isEnabled;
        },
        /**
         * @deprecated use setLogging(boolean).
         */
        set: function (enable) {
            this.logger.log('warn', 'loggingEnabled is deprecated and will be removed in a future version. Please use setLogging(boolean)');
            this.logger.enableLogging(enable);
        },
        enumerable: false,
        configurable: true
    });
    WebSocketSession.prototype.setMinLogLevel = function (level) {
        this.logger.setMinLogLevel(level);
    };
    WebSocketSession.prototype.setLogging = function (enable) {
        this.logger.enableLogging(enable);
    };
    WebSocketSession.prototype.log = function (text) {
        if (this.loggingEnabled) {
            if (this._shouldLogToServer) {
                this.logToServer(text);
            }
            else {
                this.logger.log('log', text);
            }
        }
    };
    WebSocketSession.prototype.logToServer = function (text) {
        if (this.sessionId) {
            this.sendlogMessage([text]);
        }
        else {
            this._pendingLog.push(text);
        }
    };
    WebSocketSession.prototype.sendlogMessage = function (textArray) {
        if (this._sessionId && textArray && textArray.length > 0) {
            var payload = {
                category: 'diagnostics',
                kind: 'event',
                name: 'log',
                body: { name: 'browser', text: textArray },
            };
            this.sendMessage(payload);
        }
    };
    WebSocketSession.prototype.connect = function () {
        return __awaiter(this, void 0, void 0, function () {
            var deferred;
            return __generator(this, function (_a) {
                deferred = new Deferred();
                this._closed = false;
                if (this._serverUri &&
                    (this._serverUri.startsWith('ws:') || this._serverUri.startsWith('wss:'))) {
                    // A server uri has been specified, continue with the connection
                    this.connectByWebSocket(deferred);
                }
                return [2 /*return*/, deferred.promise];
            });
        });
    };
    WebSocketSession.prototype.connectByWebSocket = function (deferred) {
        var _this = this;
        this.log("connecting to: ".concat(this._serverUri));
        if (!this._accessToken) {
            this._serverConnection = new WebSocket(this._serverUri);
        }
        else {
            this._serverConnection = new WebSocket(this._serverUri + '?access_token=' + this._accessToken);
        }
        this._serverConnection.onmessage = function (msg) {
            _this.gotMessageFromServer(msg, deferred);
        };
        this._serverConnection.onerror = function (event) {
            if (deferred.isPending()) {
                deferred.reject(makeError('websocket failed', 'serverConnectionFailed'));
            }
        };
        // wait for the websocket to open, then continue with setup
        this._serverConnection.onopen = function (event) {
            _this.log('websocket open');
            deferred.resolve();
        };
        // setup a close handler
        this._serverConnection.onclose = function (event) {
            _this.log("websocket closed: code(".concat(event.code, "), reason(").concat(event.reason, "), clean(").concat(event.wasClean, ")"));
            if (!deferred.isRejected) {
                _this.close(false, 'normal', deferred);
            }
        };
    };
    WebSocketSession.prototype.gotMessageFromServer = function (websocket_message, deferred) {
        var raw_text = websocket_message.data;
        this.log("message received: ".concat(raw_text));
        var message = JSON.parse(raw_text);
        var category = message.category;
        var name = message.name;
        var body = message.body;
        if (category !== 'webrtc') {
            // forward on non-webrtc messages (e.g. scene)
            this._onMessage(message);
            return;
        }
        if (message.kind !== 'event') {
            // currently ignore requests and responses
            return;
        }
        if (name === 'accepted') {
            this.log("accepted, session_id = ".concat(body.sessionId));
            this._sessionId = body.sessionId;
            // The session has been accepted, send any outgoing queued messages
            for (var i = 0; i < this._outgoingQueue.length; i++) {
                this._outgoingQueue[i].body.sessionId = this._sessionId;
                this.sendMessage(this._outgoingQueue[i]);
            }
            this._outgoingQueue = [];
            // Send all pending log messages
            this.sendlogMessage(this._pendingLog);
            this._pendingLog = [];
        }
        else if (name === 'close') {
            this.close(false, body.reason, deferred);
        }
    };
    WebSocketSession.prototype.sendMessage = function (message) {
        if (!this._serverConnection) {
            return;
        }
        if (this._serverConnection.readyState === WebSocket.OPEN) {
            // connected
            this._serverConnection.send(JSON.stringify(message));
        }
        else {
            this.log("not ready, discarding message: ".concat(message));
        }
    };
    WebSocketSession.prototype.close = function (sendRtcClose, reason, deferred) {
        if (sendRtcClose === void 0) { sendRtcClose = true; }
        if (reason === void 0) { reason = 'normal'; }
        if (this._closed) {
            return;
        }
        this._closed = true;
        if (deferred) {
            if (deferred.isResolved()) {
                this._onClose(reason);
            }
            else {
                deferred.reject(makeError('websocket closed: ' + reason, reason));
            }
        }
        if (this._serverConnection) {
            this.log('closing server connection');
            this._serverConnection.close();
        }
    };
    Object.defineProperty(WebSocketSession.prototype, "serverConnection", {
        get: function () {
            return this._serverConnection;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "sessionId", {
        get: function () {
            return this._sessionId;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "peerConnection", {
        get: function () {
            return null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "features", {
        get: function () {
            return this._features;
        },
        enumerable: false,
        configurable: true
    });
    WebSocketSession.prototype.sendRtcEvent = function (name, body) {
        // NOOP: Stuff for compatibility with Session in Scene
    };
    WebSocketSession.prototype.sendVideoBounds = function (widthIgnored, heightIgnored) {
        // NOOP: Stuff for compatibility with Session in Scene
    };
    WebSocketSession.prototype.sendUserText = function (text) {
        this.logger.log('error', 'WebSocketSession discarding text: ' + text);
    };
    Object.defineProperty(WebSocketSession.prototype, "microphoneMuteDelay", {
        get: function () {
            return undefined;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "microphoneMuted", {
        get: function () {
            return null;
        },
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        set: function (mute) { },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "onUserText", {
        // eslint-disable-next-line @typescript-eslint/no-empty-function
        set: function (userTextFunction) { },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "isMicrophoneConnected", {
        get: function () {
            return null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "isCameraConnected", {
        get: function () {
            return null;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "offsetX", {
        get: function () {
            return 0;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WebSocketSession.prototype, "offsetY", {
        get: function () {
            return 0;
        },
        enumerable: false,
        configurable: true
    });
    WebSocketSession.prototype.isMicrophoneActive = function () {
        return false;
    };
    WebSocketSession.prototype.isCameraActive = function () {
        return false;
    };
    WebSocketSession.prototype.setMediaDeviceActive = function (_a) {
        var microphone = _a.microphone, camera = _a.camera;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                throw makeError('setMediaDeviceActive not supported on WebSocketSession', 'notSupported');
            });
        });
    };
    return WebSocketSession;
}());
export { WebSocketSession };
